Modules/_zoneinfo.c

```diff
@@ -2243,6 +2246,95 @@ ymd_to_ord(int y, int m, int d)
     return days_before_year + yearday + d;
 }
 
+/* Convert a generic PyObject into a PyDateTime_DateTime object. */
+static PyDateTime_DateTime *
+as_datetime(PyObject *dt)
+{
+    if (PyDateTime_CheckExact(dt)) {
+        Py_INCREF(dt);
+        return (PyDateTime_DateTime *) dt;
+    }
+
+    int year, month, day;
+    int hour, minute, second;
+    unsigned char fold;
+
+    PyObject *num = PyObject_GetAttrString(dt, "year");
+    if (num == NULL) {
+        return NULL;
+    }
+    year = PyLong_AsLong(num);
+    Py_DECREF(num);
+    if (year == -1) {
```

Comment:
I think all of these checks should be for `val == -1 && PyErr_Occurred()`. Otherwise we end up returning a null value (meaning an error) without anyone having set an appropriate error string. Sending `-1` in to the pydatetime constructor is fine IIUC, since the error-checking there will raise an appropriate exception. So we just need to catch conversion errors.